pkg/gatewayserver/io/rtts.go

```diff
@@ -20,55 +20,69 @@ import (
 	"time"
 )
 
+type rttItem struct {
+	d time.Duration
+	t time.Time
+}
+
 type rtts struct {
 	count int
+	ttl   time.Duration
 	mu    sync.RWMutex
-	items []time.Duration
+	items []rttItem
 }
 
-func newRTTs(count int) *rtts {
+func newRTTs(count int, ttl time.Duration) *rtts {
 	return &rtts{
 		count: count,
-		items: make([]time.Duration, 0, count+1),
+		ttl:   ttl,
+		items: make([]rttItem, 0, count+1),
 	}
 }
 
 // Record records the given round-trip time.
-func (r *rtts) Record(d time.Duration) {
+func (r *rtts) Record(d time.Duration, t time.Time) {
 	r.mu.Lock()
-	r.items = append(r.items, d)
+	r.items = append(r.items, rttItem{d, t})
 	if len(r.items) > r.count {
-		r.items = append(r.items[:0], r.items[len(r.items)-r.count:]...)
+		r.items = append(r.items[:0], r.items[1:]...)
 	}
 	r.mu.Unlock()
 }
 
 // Last returns the last measured round-trip time.
-func (r *rtts) Last() (time.Duration, bool) {
+func (r *rtts) Last(ref time.Time) (time.Duration, bool) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	if len(r.items) == 0 {
+	if len(r.items) == 0 || ref.Sub(r.items[len(r.items)-1].t) > r.ttl {
 		return 0, false
 	}
-	return r.items[len(r.items)-1], true
+	return r.items[len(r.items)-1].d, true
 }
 
-// Stats returns the min, max, median and number of recorded round-trip times.
-func (r *rtts) Stats() (min, max, median time.Duration, count int) {
+// Stats returns the min, max, median, requested percentile and number of recorded round-trip times.
+func (r *rtts) Stats(percentile int, ref time.Time) (min, max, median, np time.Duration, count int) {
 	r.mu.RLock()
-	sorted := append(make([]time.Duration, 0, len(r.items)), r.items...)
+	sorted := make([]rttItem, 0, len(r.items))
+	for i, item := range r.items {
+		if ref.Sub(item.t) <= r.ttl {
+			sorted = append(sorted, r.items[i:]...)
+			break
+		}
+	}
 	r.mu.RUnlock()
 	if len(sorted) == 0 {
 		return
 	}
-	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
-	min = sorted[0]
-	max = sorted[len(sorted)-1]
+	sort.Slice(sorted, func(i, j int) bool { return sorted[i].d < sorted[j].d })
+	min = sorted[0].d
+	max = sorted[len(sorted)-1].d
 	if l := len(sorted); l%2 == 0 {
-		median = (sorted[l/2-1] + sorted[l/2]) / 2
+		median = (sorted[l/2-1].d + sorted[l/2].d) / 2
 	} else {
-		median = sorted[l/2]
+		median = sorted[l/2].d
 	}
+	np = sorted[int(float32(percentile)/100.0*float32(len(sorted)))-1].d
```

Comment:
Nit: flip this with the line bellow and use `count` instead of `len(sorted)`. Also does this will panic if `count` is 1 and percentile is not 100.

Reply:
You may also consider putting `count` above `min` in order to avoid creating `l`.

Reply:
> Nit: flip this with the line bellow and use `count` instead of `len(sorted)`

Doesn't matter, see https://blog.golang.org/slices-intro

> Also this will panic if `count` is 1 and percentile is not 100.

Fixed, CI caught it too fortunately